src/main/java/com/pubnub/api/models/consumer/objects_api/member/PNUUID.java

```diff
@@ -43,9 +63,14 @@ public static class UUIDWithCustom extends PNUUID {
         @JsonAdapter(CustomPayloadJsonInterceptor.class)
         private final Object custom;
 
-        UUIDWithCustom(UUIDId uuid, Object custom) {
+        private UUIDWithCustom(UUIDId uuid, Object custom) {
             super(uuid);
             this.custom = custom;
         }
+
+        private UUIDWithCustom(UUIDId uuid, Object custom, String status) {
+            super(uuid, status);
+            this.custom = custom;
+        }
```

Comment:
There's a lot of going on in this file. Here's my take. A little bit shorter

```
@Data
public abstract class PNUUID {
    @NonNull
    private final UUIDId uuid;
    private final String status;

    public static PNUUID uuid(final String uuid) {
        return new UUIDWithoutCustom(new UUIDId(uuid), null);
    }

    public static PNUUID uuid(final String uuid, final String status) {
        return new UUIDWithoutCustom(new UUIDId(uuid), status);
    }

    public static PNUUID uuidWithCustom(final String uuid, final Map<String, Object> custom) {
        return new UUIDWithCustom(new UUIDId(uuid), new HashMap<>(custom), null);
    }

    public static PNUUID uuidWithCustom(final String uuid, final Map<String, Object> custom, final String status) {
        return new UUIDWithCustom(new UUIDId(uuid), new HashMap<>(custom), status);
    }

    @Data
    public static class UUIDId {
        private final String id;
    }

    @Getter
    @EqualsAndHashCode(callSuper = true)
    @ToString
    private static class UUIDWithoutCustom extends PNUUID {
        private UUIDWithoutCustom(UUIDId uuid, String status) {
            super(uuid, status);
        }
    }

    @Getter
    @EqualsAndHashCode(callSuper = true)
    @ToString
    private static class UUIDWithCustom extends PNUUID {
        @JsonAdapter(CustomPayloadJsonInterceptor.class)
        private final Object custom;

        private UUIDWithCustom(UUIDId uuid, Object custom, String status) {
            super(uuid, status);
            this.custom = custom;
        }
    }
}
```

Reply:
Even shorter and I think works just the same (except it doesn't expose `getStatus` in PNUUID)

```
public interface PNUUID {
    @NotNull
    UUIDId getUuid();

    static PNUUID uuid(final String uuid) {
        return new UUIDWithoutCustom(new UUIDId(uuid), null);
    }

    static PNUUID uuid(final String uuid, final String status) {
        return new UUIDWithoutCustom(new UUIDId(uuid), status);
    }

    static PNUUID uuidWithCustom(final String uuid, final Map<String, Object> custom) {
        return new UUIDWithCustom(new UUIDId(uuid), new HashMap<>(custom), null);
    }

    static PNUUID uuidWithCustom(final String uuid, final Map<String, Object> custom, final String status) {
        return new UUIDWithCustom(new UUIDId(uuid), new HashMap<>(custom), status);
    }

    @Data
    class UUIDId {
        private final String id;
    }

    @Data
    @RequiredArgsConstructor(access = AccessLevel.PRIVATE)
    class UUIDWithoutCustom implements PNUUID {
        @NonNull
        private final UUIDId uuid;
        private final String status;
    }

    @Data
    @RequiredArgsConstructor(access = AccessLevel.PRIVATE)
    class UUIDWithCustom implements PNUUID {
        @NonNull
        private final UUIDId uuid;
        @JsonAdapter(CustomPayloadJsonInterceptor.class)
        private final Object custom;
        private final String status;
    }
}
```

Reply:
I am not used to have interface with some much stuff like inner classes, static methods. Looks strange to me but I can give it a try

Reply:
What's the difference between interface with lots of stuff inside and abstract class with lots of stuff inside? :)

Reply:
It think my reason is that I am not use to. I was leaving with this statement in mind:
"Ideally, interfaces shouldn't encapsulate behaviour, and we should only use them for defining the public API of a certain type."   
but since java provide possibility to create static methods in Interface why not to use them.